protocol/test/amo/oud/oud-token.ts

```diff
@@ -0,0 +1,210 @@
+import { expect } from "chai";
+import { OudToken, OudToken__factory } from "../../../typechain";
+// import { recoverToken, shouldRevertNotOwner, testErc20Permit } from "../../helpers";
+import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
+import { BigNumberish, Signer, TypedDataDomain, TypedDataField } from "ethers";
+import { ethers } from "hardhat";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import { splitSignature } from "ethers/lib/utils";
+import { blockTimestamp } from "../../helpers";
+ 
+describe("Mintable Token", async () => {
+    let token: OudToken;
+    let owner: SignerWithAddress;
+    let minter: SignerWithAddress;
+    let alan: SignerWithAddress;
+    let spender: SignerWithAddress;
+
+    before(async () => {
+        [owner, minter, alan, spender] = await ethers.getSigners();
+    });
+
+    async function setup() {
+        token = await new OudToken__factory(owner).deploy("Oud Test Token", "Oud");
+
+        return {
+            token,
+        };
+    }
+
+    beforeEach(async () => {
+        ({
+            token,
+        } = await loadFixture(setup));
+    });
+
+    it("Init", async () => {
+        expect(await token.symbol()).eq("Oud");
+        expect(await token.name()).eq("Oud Test Token");
+        expect(await token.totalSupply()).eq(0);
+    });
+
+    it("admin", async () => {
+        await expect(token.connect(alan).recoverToken(token.address, alan.getAddress(), 10)).to.revertedWith("Ownable: caller is not the owner");
+
+        // Happy paths
+        await expect(token.recoverToken(token.address, alan.getAddress(), 10))
+            .to.revertedWith("ERC20: transfer amount exceeds balance");
+    });
+
+    it("owner can recover tokens", async () => {
+        const amount = 50;
+        await token.addMinter(owner.getAddress());
+        await token.mint(token.address, amount);
+        // await recoverToken(token, amount, token, owner);
```

Comment:
```

export async function recoverToken(token: Contract, amount: BigNumberish, from: Contract, signer: Signer) {
  const balBefore = await token.balanceOf(await signer.getAddress());
  const lib = CommonEventsAndErrors__factory.connect(from.address, signer);
  await expect(from.recoverToken(token.address, await signer.getAddress(), amount))
      .to.emit(lib, "TokenRecovered")
      .withArgs(await signer.getAddress(), token.address, amount);
  const balAfter = await token.balanceOf(await signer.getAddress());
  expect(balAfter.sub(balBefore)).eq(amount);
}
```

Reply:
fixed in [3077781](https://github.com/TempleDAO/temple/pull/760/commits/3077781b104e81a77a9ffe07e54b7f39e4bcf27f)